Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -1509,6 +1545,138 @@ DiscusMultipleScatteringCorrection::scatter(const int nScatters, Kernel::PseudoR
   return {true, weights};
 }
 
+/*
+ * Construct a hexahedron shape extending from the detector's front face across the collimator openning window toward
+ the sample with a legth as twice the distance from sample to detector.
+ */
+std::shared_ptr<Geometry::CSGObject> DiscusMultipleScatteringCorrection::createCollimatorHexahedronShape(
+    const Kernel::V3D &samplePos, const Mantid::Geometry::DetectorInfo &detectorInfo, const int64_t &histogramIndex) {
+  const auto shape = detectorInfo.detector(histogramIndex).shape();
+  try {
+    shape->shapeInfo();
+  } catch (std::exception &) {
+    return nullptr;
+  }
+
+  const auto &detectorId = detectorInfo.detector(histogramIndex).getID();
+  const auto &detectorAbsolPos = detectorInfo.position(detectorInfo.indexOf(detectorId));
+  const auto cuboidGeometry = shape->shapeInfo().cuboidGeometry();
+
+  // Positions of the detector's front face
+  const auto &detLeftFrontBottomPos = detectorAbsolPos + cuboidGeometry.leftFrontBottom;
+  const auto &detLeftFrontTopPos = detectorAbsolPos + cuboidGeometry.leftFrontTop;
+  const auto &detRightFrontBottomPos = detectorAbsolPos + cuboidGeometry.rightFrontBottom;
+  const auto detRightFrontTopPos = detLeftFrontTopPos + detRightFrontBottomPos - detLeftFrontBottomPos;
+
+  const auto detCentorPos =
+      (detLeftFrontBottomPos + detLeftFrontTopPos + detRightFrontBottomPos + detRightFrontTopPos) / 4.0;
+  const auto detTopMiddlePos = (detLeftFrontTopPos + detRightFrontTopPos) / 2.0;
+
+  // Sanity check to avoid dividing by zero
+  if ((detRightFrontTopPos == detLeftFrontTopPos) || (detTopMiddlePos == detCentorPos) || (detCentorPos == samplePos)) {
+    return nullptr;
+  }
+
+  // Define the unit vectors needed for calculations
+  const auto unitVecLeftToRight =
+      (detRightFrontTopPos - detLeftFrontTopPos) / (detRightFrontTopPos - detLeftFrontTopPos).norm();
+  const auto unitVecBottomToUp = (detTopMiddlePos - detCentorPos) / (detTopMiddlePos - detCentorPos).norm();
```

Comment:
I'm not sure about this - will this still work for both possible configurations of collimator we know about (i.e. axis of cylinder vertical, and axis of cylinder along beam)?

Reply:
I think it should work, the names like *Up given for the variable are just to tally with the paper drawing I had and they should work with vector geometry irrespective of the coordinate system.